Refactor optimization tutorial (#161)
micro-ROS/micro-ROS.github.io: _docs/tutorials/core/microxrcedds_rmw_configuration/index.md

permalink: /docs/tutorials/core/microxrcedds_rmw_configuration/
---

> micro-ROS targets microcontroller, devices with low memory resources which normally are use in industrial or production environments where safety is one of the main requirements.

Reviewer: As micro-ROS is not developed according to any safety standard, please do not argue with safety here but only with the low memory resources.

For example, most of the safety-critical embedded software certifications do not allow the use of dynamic memory allocation after the initialization phase.
Therefore, in such as scenarios, memory management is a keep concept.

With that in mind, micro-ROS try to address the memory management issue prioritizing the use of static memory instead of dynamic memory and optimizing the memory footprint of the applications. This, of course, has a cost that the users must agree to pay, a precompile tunning.

> This tutorial shows which are the memory resources managed by micro-ROS and how to tuning them for a particular application.

Reviewer: ```suggestion
This tutorial explains which are the memory resources managed by micro-ROS and how to tune them for a particular application.
```

## Memory resources

### Micro XRCE-DDS

Micro XRCE-DDS messages flows between Client and Agent throw **streams**.

> A stream represents an independent ordered flow of information, that  is, it is a sort of messaging queue.

Reviewer: ```suggestion
A stream represents an independent ordered flow of information, that is, it is a sort of messaging queue.
```

The use of best-effort or reliable stream is handled by the `rmw_qos_reliability_policy_t` set in the `rmw_qos_profile_t` for a particular publisher or subscription.
In the case of `RMW_QOS_POLICY_RELIABILITY_BEST_EFFORT` best-effort streams are used and for `RMW_QOS_POLICY_RELIABILITY_RELIABLE` reliable streams.

### rmw-microxrcedds

The RMW implementation for Micro XRCE-DDS (rmw-microxrcedds) uses static memory for allocating the resources associated with the `rcl` and `rclc` entities.
This static memory is managed thanks to independent memory pools for each kind of entity.
The size of these memory pools could be set through CMake flags,
for example, the `RMW_UXRCE_MAX_PUBLISHERS` sets the size of the `rcl_publisher_t`'s pool memory.
It should be noted that the size of these memory pools restricts the maximum number of entities that a micro-ROS application could use.

The figure below summarizes the relation between the `rcl` entities and the CMake flags, which the size of the memory pool associated with such entity.

![](./imgs/micro_ros_memory.svg)

> Another important memory resource managed by the rmw-microxrce is the message history.

Reviewer: ```suggestion
Another important memory resource managed by the rmw-microxrcedds is the message history.
```
